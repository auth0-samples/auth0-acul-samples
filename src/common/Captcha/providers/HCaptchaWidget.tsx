import React, { useEffect, useState } from "react";
import type { CaptchaWidgetProps, CaptchaResponse } from "../index";
import { cn } from "@/utils/helpers/cn";

const HCaptchaWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  className = "",
}) => {
  const [token, setToken] = useState("");

  if (config.provider !== "hcaptcha") {
    return null;
  }

  const siteKey = config.siteKey;

  useEffect(() => {
    const addWidgetScript = document.createElement("script");
    addWidgetScript.src = "https://js.hcaptcha.com/1/api.js";
    addWidgetScript.async = true;
    addWidgetScript.defer = true;
    document.body.appendChild(addWidgetScript);

    (window as any).updateToken = (token: string) => {
      setToken(token);
      const response: CaptchaResponse = {
        provider: "hcaptcha",
        token: token,
      };
      onCaptchaResponse(response);
    };

    return () => {
      if (document.body.contains(addWidgetScript)) {
        document.body.removeChild(addWidgetScript);
      }
      // Clean up global callback
      delete (window as any).updateToken;
    };
  }, [onCaptchaResponse]);

  return (
    <div className={cn("space-y-2", className)}>
      <div className="flex justify-center w-full">
        <div
          id="h-captcha"
          className="h-captcha"
          data-sitekey={siteKey}
          data-callback="updateToken"
          data-theme="light"
          data-size="normal"
        />
        <input
          type="hidden"
          name="captcha"
          id="hidden-h-captcha"
          value={token}
        />
      </div>
    </div>
  );
};

export default HCaptchaWidget;
