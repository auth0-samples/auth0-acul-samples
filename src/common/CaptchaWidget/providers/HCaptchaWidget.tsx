import React, { useEffect, useRef, useState } from "react";
import type { CaptchaWidgetProps, CaptchaResponse } from "../types";

declare global {
  interface Window {
    hcaptcha: any;
  }
}

const HCaptchaWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  onError,
  className = "",
}) => {
  const hcaptchaRef = useRef<HTMLDivElement>(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [widgetId, setWidgetId] = useState<string | null>(null);

  if (config.provider !== "hcaptcha") {
    return null;
  }

  const siteKey = config.siteKey;
  const theme = config.theme || "light";
  const size = config.size || "normal";

  useEffect(() => {
    // Load hCaptcha script
    const loadHCaptcha = () => {
      if (window.hcaptcha) {
        setIsLoaded(true);
        return;
      }

      const script = document.createElement("script");
      script.src =
        "https://js.hcaptcha.com/1/api.js?onload=onHCaptchaLoad&render=explicit";
      script.async = true;
      script.defer = true;

      // Global callback for when hCaptcha loads
      (window as any).onHCaptchaLoad = () => {
        setIsLoaded(true);
      };

      document.head.appendChild(script);
    };

    loadHCaptcha();

    return () => {
      // Cleanup
      if (widgetId && window.hcaptcha) {
        try {
          window.hcaptcha.remove(widgetId);
        } catch (e) {
          console.warn("Error removing hCaptcha:", e);
        }
      }
    };
  }, []);

  useEffect(() => {
    if (isLoaded && hcaptchaRef.current && !widgetId) {
      try {
        const id = window.hcaptcha.render(hcaptchaRef.current, {
          sitekey: siteKey,
          theme: theme,
          size: size,
          callback: (token: string) => {
            const response: CaptchaResponse = {
              provider: "hcaptcha",
              token,
            };
            onCaptchaResponse(response);
          },
          "expired-callback": () => {
            onCaptchaResponse(null);
          },
          "error-callback": () => {
            onError?.("hCaptcha error occurred");
            onCaptchaResponse(null);
          },
        });
        setWidgetId(id);
      } catch (error) {
        console.error("Error rendering hCaptcha:", error);
        onError?.("Failed to render hCaptcha");
      }
    }
  }, [isLoaded, siteKey, theme, size, onCaptchaResponse, onError, widgetId]);

  return (
    <div className={`hcaptcha-container ${className}`}>
      <div ref={hcaptchaRef}></div>
    </div>
  );
};

export default HCaptchaWidget;
