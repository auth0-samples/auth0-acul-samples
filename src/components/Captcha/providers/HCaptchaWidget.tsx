import React, { useCallback, useEffect, useState } from "react";

import HCaptcha from "@hcaptcha/react-hcaptcha";

import { cn } from "@/lib/utils";

import type { CaptchaResponse, CaptchaWidgetProps } from "../index";

const HCAPTCHA_PROVIDER = "hcaptcha";

const HCaptchaWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  className = "",
  error,
  theme,
}) => {
  const [token, setToken] = useState<string | undefined>(undefined);
  const [errorMessage, setErrorMessage] = useState<string | undefined>(error);
  const [isClient, setIsClient] = useState(false);

  console.log("HCaptchaWidget errorMessage:", errorMessage);

  const hCaptchaWidgetStyle = {
    transform: "scale(1.06)",
    transformOrigin: "0 0",
  };

  const handleVerify = useCallback(
    (value: string | null) => {
      const newToken = value || undefined;
      setToken(newToken);
      setErrorMessage(undefined);

      if (value) {
        const response: CaptchaResponse = {
          provider: HCAPTCHA_PROVIDER,
          token: value,
        };
        onCaptchaResponse(response);
      } else {
        onCaptchaResponse(null);
      }
    },
    [onCaptchaResponse]
  );

  const handleExpired = useCallback(() => {
    handleVerify(null);
  }, [handleVerify]);

  const handleError = useCallback(() => {
    setToken(undefined);
    onCaptchaResponse(null);
  }, [onCaptchaResponse]);

  useEffect(() => {
    // This effect runs only on the client side after the component mounts
    setIsClient(true);
  }, []);

  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  if (config.provider !== HCAPTCHA_PROVIDER) {
    return null;
  }
  const { siteKey } = config;
  if (!siteKey) {
    return null;
  }

  const getHCaptchaTheme = (): "light" | "dark" => {
    if (
      isClient &&
      theme === "auto" &&
      window.matchMedia("(prefers-color-scheme: dark)").matches
    ) {
      return "dark";
    }
    return theme === "dark" ? "dark" : "light";
  };

  // Only render HCaptcha on the client side
  if (!isClient) {
    return null;
  }

  return (
    <div className={cn("space-y-2", className)}>
      <div className="flex justify-center w-full">
        <div style={hCaptchaWidgetStyle}>
          <HCaptcha
            sitekey={siteKey}
            onVerify={handleVerify}
            onExpire={handleExpired}
            onError={handleError}
            theme={getHCaptchaTheme()}
            size="normal"
          />
        </div>
        <input
          type="hidden"
          name="h-captcha-response"
          id="hidden-h-captcha"
          value={token || ""}
        />
      </div>
    </div>
  );
};

export default HCaptchaWidget;
