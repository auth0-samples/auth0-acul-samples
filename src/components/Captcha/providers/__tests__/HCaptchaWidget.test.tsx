import { FormProvider, useForm } from "react-hook-form";

import { render, screen } from "@testing-library/react";

// Mock HCaptcha so it renders a testable element
jest.mock("@hcaptcha/react-hcaptcha", () => ({
  __esModule: true,
  default: (props: any) => <div data-testid="hcaptcha" {...props} />,
}));

import HCaptchaWidget from "../HCaptchaWidget";

const baseProps = {
  config: {
    provider: "hcaptcha",
    siteKey: "test-site-key",
  },
  onCaptchaResponse: jest.fn(),
  className: "",
  theme: "light" as const,
  error: undefined,
  control: undefined,
  name: "captcha",
};

function Wrapper({
  children,
  defaultValues = {},
}: {
  children: any;
  defaultValues?: any;
}) {
  const methods = useForm({ defaultValues });
  return (
    <FormProvider {...methods}>
      <form>{children({ control: methods.control })}</form>
    </FormProvider>
  );
}

describe("HCaptchaWidget", () => {
  it("renders the HCaptcha widget", () => {
    render(<Wrapper>{() => <HCaptchaWidget {...baseProps} />}</Wrapper>);
    expect(screen.getByTestId("hcaptcha")).toBeInTheDocument();
  });

  it("matches snapshot (default)", () => {
    const { container } = render(
      <Wrapper>{() => <HCaptchaWidget {...baseProps} />}</Wrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot (with error)", () => {
    const { container } = render(
      <Wrapper>
        {() => <HCaptchaWidget {...baseProps} error="Test error" />}
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
