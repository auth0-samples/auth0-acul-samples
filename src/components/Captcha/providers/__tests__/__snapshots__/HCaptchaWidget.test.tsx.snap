// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HCaptchaWidget matches snapshot (default) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="flex justify-center w-full"
      >
        <div
          style="transform: scale(1.06); transform-origin: 0 0;"
        >
          <div
            data-testid="hcaptcha"
            sitekey="test-site-key"
            theme="light"
          />
        </div>
        <input
          id="hidden-h-captcha"
          name="h-captcha-response"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`HCaptchaWidget matches snapshot (with error) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="flex justify-center w-full"
      >
        <div
          style="transform: scale(1.06); transform-origin: 0 0;"
        >
          <div
            data-testid="hcaptcha"
            sitekey="test-site-key"
            theme="light"
          />
        </div>
        <input
          id="hidden-h-captcha"
          name="h-captcha-response"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;
